src/main/java/frc/robot/gamepieces/AbstractLayers/ShooterAL.java

```diff
@@ -0,0 +1,375 @@
+          hoodFollower = new Servo(RobotMap.HOOD_FOLLOWER_PWM_PORT);
+        }
+
+      } else {
+        hoodLeader = null;
+        hoodFollower = null;
+      }
+
+      if (RobotMap.HAS_SHOOTER_LEDS) {
+        leds = new AddressableLED(RobotMap.SHOOTER_LED_CHANNEL);
+        ledBuffer = new AddressableLEDBuffer(RobotMap.SHOOTER_LED_AMOUNT * (RobotMap.SHOOTER_DOUBLESIDE_LED? 2: 1));
+        leds.setLength(ledBuffer.getLength());
+
+        for (var i = 0; i < ledBuffer.getLength(); i++) {
+          ledBuffer.setRGB(i, 0, 0, 0);
+        }
+
+        leds.setData(ledBuffer);
+        leds.start();
+      } else {
+        leds = null;
+        ledBuffer = null;
+      }
+
+      instance = new ShooterAL(flywheel);
+      instance.stop();
+    }
+    return instance;
+  }
+
+
+  private ShooterAL(TalonSpeedControllerGroup flywheel) {
+    super("Telemetry", "Shooter");
+  }
+
+  public TalonSpeedControllerGroup getMotor() {
+    return flywheel;
+  }
+
+  public void stop() {
+    //LOGGER.error("Stop called");
+    flywheel.stopMotor();
+  }
+
+  public void setSpeed(double speed) {
+    this.speed = speed;
+    
+  }
+
+  public void rampToSpeed(double speed) {
+    if (flywheel != null && RobotMap.HAS_SHOOTER) {
+      double output = Math.max(-1.0, Math.min(1.0, speed));
+      flywheel.set(ControlMode.Velocity, output);
+      LOGGER.debug("the speed is {}", output);
+    }
+  }
+
+  public boolean atSpeed() {
+    double current = flywheel.velocity();
+    double target = flywheel.closedLoopTarget();
+    double error = current / target;
+
+    if (error <= 1 + RobotMap.SHOOTER_SPEED_TOLERANCE || error >= 1 - RobotMap.SHOOTER_SPEED_TOLERANCE) {
+      return true;
+    } else {
+      return false;
+    }
+  }
+
+
+  public void startShooting() {
+    if (trigger != null && RobotMap.HAS_TRIGGER) {
+      trigger.set(1.0);
+      LOGGER.debug(trigger.get());
+    }
+  }
+
+  public void stopShooting() {
+    if (trigger != null && RobotMap.HAS_TRIGGER) {
+      trigger.set(0.0);
+    }
+  }
+
+  public void reverseShooter() {
+    if (trigger != null && RobotMap.HAS_TRIGGER) {
+      trigger.set(-1.0);
+    }
+  }
+
+  public void setTriggerState(boolean state) {
+    if (trigger != null && RobotMap.HAS_TRIGGER) {
+      this.triggerState = state;
+    }
+  }
+
+  public void flyWheelPIDF(double kP, double kI, double kD, double kF, double kMaxVelocity) {
+    if (flywheel != null && RobotMap.HAS_SHOOTER) {
+      flywheel.pidf(RobotMap.SHOOTER_PID_SLOT_DRIVE, kP, kI, kD, kF, kMaxVelocity);
+    }
+  }
+
+  public void setLedSrip(int r, int g, int b, int startingLed, int endingLed) {
+    if (ledBuffer != null && leds != null && RobotMap.HAS_SHOOTERLEDS) {
+      for (var i = Math.max(0, startingLed); i <= Math.min(ledBuffer.getLength()-1, endingLed); i++) {
+        ledBuffer.setRGB(i, r, g, b);
+     }
+     leds.setData(ledBuffer);
+    }
+  }
+
+  public void fillStrip(int r, int g, int b, int led) {
```

Comment:
All of these LED functions need to be moved to another class. They do not belong in the shooter class.